packages/experimental-utils/src/ast-utils/predicates.ts

```diff
@@ -1,5 +1,12 @@
 import { AST_NODE_TYPES, AST_TOKEN_TYPES, TSESTree } from '../ts-estree';
 
+const isNodeOfType =
+  <NodeType extends AST_NODE_TYPES>(nodeType: NodeType) =>
+  (
+    node: TSESTree.Node | null | undefined,
+  ): node is TSESTree.Node & { type: NodeType } =>
```

Comment:
I'll be honest when I say that I don't know how TS resolves this exactly.

It seems wrong that `(x | y | z) & a` extracts the type from the union that is assignable to `a` because the operators are commutative (i.e. the above is supposed to be the same as `x & a | y & a | z & a`).

TS must have some magic to make this as it does really do some weird shit:
![image](https://user-images.githubusercontent.com/7462525/127786359-954beda8-c2ed-4e5a-bf8a-73e36517c5f0.png)

[It only seems to work this way for discriminated union types.](https://www.typescriptlang.org/play?#code/C4TwDgpgBAQghgZ2gXigb1JAXFA5HXAGijh11wF8oAfdTCMgIyKkZwEYKBuAKHqgAKAJwD2YKKnhIoAMnSk8lXvwAiASwQBjAKoA7NSN0TYiaHIzgGeAtx49NhhMChhRYHMLHGL2PM2IK5MRsUJxcUAD0EVCAoOT2js4AJho6+oY46lp6BkaoPlb4LCFhkdGAMuQ8QA)

Reply:
It's normal that TS is narrowing down the type of the Union.

You start at a wide `Node` type (which all have a unique `type`) and narrow it down by doing the extra `& { type: AST_NODE_TYPES.VariableDeclarator }`.
This is because all the other types will be `{ type: A } & { type: B }`, which will resolve in `never` as that's never possible.

So it resolves to `TSEstree.VariableDeclarator & { type: AST_NODE_TYPES.VariableDeclarator }`, which is the same as just `TSEstree.VariableDeclarator`

Reply:
Ah.. I didn't think about the fact that `'a' & 'b' === never`.
That makes sense then.

Reply:
Wait still not there.

Why does it not just `never` out the `type` property?

If I do `{a: string} & {b: string}` you end up with `{a: string, b: string}`, so why does `{type: 'a', a: string} & {type: 'b'}` not compute to `{type: never, a: string}`?

Reply:
That indeed sounds a bit like an inconsistency.

But on the other hand it's logical as you're actually merging 2 object with each other and you can't merge them as `type` is something else? 🤔

Reply:
Maybe @orta can give more insights into this?

I'm just thinking out loud here (not really sure why TS works this way tbh, I just know that it works 😛)

Reply:
As another example.
```ts
type T1 = {a: string, b: number} & {a: number, c: string};
// typeof T1 === {a: never, b: number, c: string}

declare const n: never;
const x: T1 = { a: n, b: 1, c: '' }; // no TS error
const y: T1 = { b: 1, c: '' }; // Error: Property 'a' is missing
```
[repl](https://www.typescriptlang.org/play?#code/C4TwDgpgBAKgjFAvFA3gQwFxQM7AE4CWAdgOYA0UARlkQK4C2lEeAvlAGSqZR2PMUBjLLkKkWAbgBQkgCYQBAGzR5oAgPZFcPGhABuzKes3AoADyzwkqKNyIVqUOIKwByF1AmSjWkBYTIUKiwnKCEoNw8pIA)

So it seems like there is literally just special handling for string literal types?

Reply:
We either discovered a bug **or** we're missing something here I think.
There must be a reason why this is different then, no?

Reply:
But having a unique thing is the way to get the sub-type of a certain union, so I think we're actually correct in what I'm doing here.

```ts
type A = { a: 'A' };
type B = { b: 'B' };
type C = A | B;
```

To know if it's of type `A`, you check for something that's unique to `A`, which is the `a` key.

With `Node`, the `type` key is the unique distinguisher.

Reply:
Twitter thread: https://twitter.com/bradzacher/status/1421954710592921600